### PR TITLE
Fix broken category sub category section on filtering

### PR DESCRIPTION
**Pull request by galund**
Categories tree 'selected filters' must include parent category. This ensures the parent category (implied) selection survives when a filter is selected - we get a hidden field in the filters form.

https://trello.com/c/QPkflfgt/122-broken-category-sub-category-section-on-filtering-issue-exists-on-production

### CODE REVIEW

File: app/main/presenters/search_presenters.py [hunk 1]
```diff
@@ -288,8 +288,18 @@ def _annotate_categories_with_selection(lot_slug, category_filters, request, url
             content_manifest, framework, aggregations, keys_to_remove, parent_category=category['value'])
 
         if selected_descendants:
-            # If a parentCategory has been sent as a url query param, de-select all other parent categories.
-            if request.values.get('parentCategory', category['value']) != category['value']:
+            # If a parentCategory has been sent as a url query param, and it's this category...
+            # (Note: `get` fallback is for if there are selected descendants but no parent category is set, which is
+            # not expected, but could happen if constructing URLs by hand.)
+            if request.values.get('parentCategory', category['value']) == category['value']:
```
Comment (risicle) [inline app/main/presenters/search_presenters.py]: I'm *really* confused over werkzeug's differentiation between `args` `form` and `values` and it seems from a distance to be one of the worst designed things ever.

Comment (galund) [reply app/main/presenters/search_presenters.py]: Me too. So I just looked it up: args is from the GET query string; form is from the POST; values is a combination of the two, e.g. for if your form might be a POST or a GET and you don't much know or care.

I suppose in these presenters, `values` is the right thing, because they don't know whether the view that called them was a POST or a GET. Really, they probably shouldn't be poking around in the request, so the selected parent category should have been passed in. But this was easier at the time, and now doesn't seem like the moment to revisit whether our search presenters are sufficiently decoupled...

Comment (risicle) [reply app/main/presenters/search_presenters.py]: Yeah, the thing I can't figure out is what you're supposed to do if you want to access the POSTed parameters unsullied by any query strings. Anyway, not that relevant here...

Comment (galund) [reply app/main/presenters/search_presenters.py]: That's `request.form` - I think. In other words, that's not an actual form object (because that would introduce a dependency on wtforms or whatever, and flask doesn't do that) - it's just another multidict.

Comment (risicle) [reply app/main/presenters/search_presenters.py]: Well I would think that if the docs didn't do on to say

```
Changed in version 0.9: Previous to Werkzeug 0.9 this would only contain form data for POST and PUT requests.
```

Which I take to mean pulls in the parameters from GET requests...?

Comment (galund) [reply app/main/presenters/search_presenters.py]: For the record, we've established it won't look in the URL/GET params, but regardless of HTTP method it will look in the request body, as long as the content type is `multipart/form-data` or equivalent.

File: app/main/presenters/search_presenters.py [hunk 2]
```diff
@@ -288,8 +288,18 @@ def _annotate_categories_with_selection(lot_slug, category_filters, request, url
...
+                # ... then ensure this choice survives as a hidden field for the filters form
+                parent_category_filter = dict()
+                parent_category_filter['name'] = 'parentCategory'
+                parent_category_filter['value'] = category['value']
```
Comment (katstevens) [inline app/main/presenters/search_presenters.py]: I might be missing something obvious here but is there any reason why this isn't written on one line as 
```
parent_category_filter = {'name': 'parentCategory', 'value': category['value']}
```
?

File: tests/main/presenters/test_search_presenters.py [hunk 3]
```diff
@@ -256,7 +256,7 @@ def test_show_lots_and_categories_selection(self, search_api_client):
         with self.app.test_request_context(url):
```
Comment (katstevens) [inline tests/main/presenters/test_search_presenters.py]: This long test feels like it could be split in two (with maybe some shared setup/context)?

File: app/main/presenters/search_presenters.py [hunk 4]
```diff
@@ -288,8 +288,18 @@ def _annotate_categories_with_selection(lot_slug, category_filters, request, url
             content_manifest, framework, aggregations, keys_to_remove, parent_category=category['value'])
 
         if selected_descendants:
-            # If a parentCategory has been sent as a url query param, de-select all other parent categories.
-            if request.values.get('parentCategory', category['value']) != category['value']:
+            # If a parentCategory has been sent as a url query param, and it's this category...
+            # (Note: `get` fallback is for if there are selected descendants but no parent category is set, which is
+            # not expected, but could happen if constructing URLs by hand.)
```
Comment (katstevens) [inline app/main/presenters/search_presenters.py]: Is the fallback scenario unlikely enough that we don't need a test for it? Or is already covered somewhere?

Comment (galund) [reply app/main/presenters/search_presenters.py]: The fallback was already in place: I was just documenting it, as it took me a while to understand why it was there. No, I don't think we test for it - it feels like an edge case, and seeing as it can't happen (I think!), it's probably okay not to test it.

The tests already run both these code paths, anyway, regardless of that fallback.

File: app/main/presenters/search_presenters.py (hunk 2 above)
Comment (galund) [reply app/main/presenters/search_presenters.py]: Point taken, this was just done in the same style as where we build other "filter dictionaries" - see `search_presenters.build_lots_and_categories_link_tree` for example. Some of those cases it's perhaps more justifiable to use this format as some of the values are the result of relatively long function calls (e.g. see the `root_node['link'] = search_link_builder(...)` line).

I think if I change this I should change them all, and I'm not sure the end result would necessarily be more readable?

File: app/main/presenters/search_presenters.py (hunk 1 above)
Comment (risicle) [reply app/main/presenters/search_presenters.py]: oh and this is probably clearer if written:

```
if 'parentCategory' not in request.values or request.values.get('parentCategory') == category['value']:
```

File: app/main/presenters/search_presenters.py (hunk 2 above)
Comment (risicle) [reply app/main/presenters/search_presenters.py]: Or even just

```
selected_category_filters.append({
    'name': 'parentCategory',
    'value': category['value'],
})
```

Comment (katstevens) [reply app/main/presenters/search_presenters.py]: Hmm. Consistency probably wins in this case? Your call @galund...

File: app/main/presenters/search_presenters.py (hunk 1 above)
Comment (galund) [reply app/main/presenters/search_presenters.py]: You're right that the actual logic becomes more clear, but it calls out the weird case even more, which doesn't make the overall 'normal' flow more clear? Also I'd have to rewrite the comment, which already made my head hurt enough...

File: app/main/presenters/search_presenters.py (hunk 2 above)
Comment (galund) [reply app/main/presenters/search_presenters.py]: Going with consistency for now if that's ok. I'll bear in mind that ppl prefer `{..}` notation next time I'm writing this kind of thing :)

File: app/main/presenters/search_presenters.py (hunk 1 above)
Comment (risicle) [reply app/main/presenters/search_presenters.py]: Weird cases are just as important as the "normal" case to me but whatevs
